### PR TITLE
feat: add support for BIP-388 wallet policies

### PR DESCRIPTION
**Pull request by trevarj**
Implemented BIP-388 by introducing a new type, `WalletPolicy`, which can be used
to create descriptor templates.

The idea is pretty simple and only really required making a new type that
implements `MiniscriptKey`, then the existing parser and `Translator` trait
takes care of the rest.

There are a few edge cases that require validation, which isn't so nice, but
works for now. Test cases that require BIP-390 and BIP-387's sortedmulti_a are
commented out and issues could arise with those test vectors when/if those BIPs
are implemented.

See `WalletPolicy`'s doc and the unit tests for usage.

### CODE REVIEW

**Comment by trevarj** [comment]
Closes #899

**Comment by tcharding** [comment]
API design suggestion but caveat this is not my repo so feel free to ignore me.

- Would it be better to put validation inside `from_descriptor`, have `from_descriptor_unchecked` (if needed)?
- Perhaps not even have the `TryFrom`s? Or at least put the logic in the type methods and in the `TryFrom` just call the methods?

As far as semantics, it looks like the unit test prove the PR sufficiently to me.

File: src/descriptor/wallet_policy/mod.rs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+/// // Convert from a Descriptor<DescriptorPublicKey> string:
+/// let policy2 = WalletPolicy::from_str(desc_str).unwrap();
+/// assert_eq!(policy1, policy2);
+/// // Convert from/to a wallet policy template string:
+/// let from_template = WalletPolicy::from_str("pkh(@0/**)").unwrap();
+/// assert_eq!(from_template.to_string(), "pkh(@0/**)");
+/// // Cannot go back into descriptor if you created from template:
+/// assert!(from_template.into_descriptor().is_err());
+/// // Convert into a full descriptor:
+/// assert_eq!(policy1.into_descriptor().unwrap(), descriptor);
```
[the remark this answers is not in the document]
Comment (trevarj) [reply src/descriptor/wallet_policy/mod.rs]: You're absolutely right. Fixed it.

If you're still using Emacs, you could switch to `rust-ts-mode` for nicer highlighting. Have to check if your build supports treesitter or not.

```elisp
(use-package rust-mode
  :custom
  (rust-mode-treesitter-derive (treesit-available-p)))
```

**Comment by trevarj** [comment]
@tcharding 

> Trevor!! 

TOBIN! Thank you for the review

> Can this be split up a bit mate? The wildcard display stuff, then maybe the `XKeyParseError` thing, then the rest?

Split into four separate commits.


> * Would it be better to put validation inside `from_descriptor`, have `from_descriptor_unchecked` (if needed)?

Seems better to me. Easier to track down the logic than having to find the TryFrom. 

> Or at least put the logic in the type methods and in the `TryFrom` just call the methods?

Done

**Comment by trevarj** [comment]
> ACK [945adb7](https://github.com/rust-bitcoin/rust-miniscript/commit/945adb7e60d2f3713123489b14156c06a81a47ab); successfully ran local tests; looks great!

Thanks for reviewing! 
Looking forward to using this for hardware wallet descriptor registration.
